homeassistant/components/tplink/sensor.py

```diff
@@ -1,182 +1,79 @@
-        device_class=SensorDeviceClass.CURRENT,
-        state_class=SensorStateClass.MEASUREMENT,
-        emeter_attr="current",
-        precision=2,
-    ),
-)
-
-
-def async_emeter_from_device(
-    device: Device, description: TPLinkSensorEntityDescription
-) -> float | None:
-    """Map a sensor key to the device attribute."""
-    if attr := description.emeter_attr:
-        if (val := getattr(device.emeter_realtime, attr)) is None:
-            return None
-        return round(cast(float, val), description.precision)
-
-    # ATTR_TODAY_ENERGY_KWH
-    if (emeter_today := device.emeter_today) is not None:
-        return round(cast(float, emeter_today), description.precision)
-    # today's consumption not available, when device was off all the day
-    # bulb's do not report this information, so filter it out
-    return None if device.device_type == DeviceType.Bulb else 0.0
-
-
-def _async_sensors_for_device(
-    device: Device,
-    coordinator: TPLinkDataUpdateCoordinator,
-    parent: Device | None = None,
-) -> list[SmartPlugSensor]:
-    """Generate the sensors for the device."""
-    return [
-        SmartPlugSensor(device, coordinator, description, parent)
-        for description in ENERGY_SENSORS
-        if async_emeter_from_device(device, description) is not None
-    ]
-
-
 async def async_setup_entry(
     hass: HomeAssistant,
-    config_entry: ConfigEntry,
+    config_entry: TPLinkConfigEntry,
     async_add_entities: AddEntitiesCallback,
 ) -> None:
     """Set up sensors."""
-    data: TPLinkData = hass.data[DOMAIN][config_entry.entry_id]
+    data = config_entry.runtime_data
     parent_coordinator = data.parent_coordinator
     children_coordinators = data.children_coordinators
-    entities: list[SmartPlugSensor] = []
-    parent = parent_coordinator.device
-    if not parent.has_emeter:
-        return
-
-    if parent.children:
-        # Historically we only add the children if the device is a strip
-        for idx, child in enumerate(parent.children):
-            # Only iot strips have child coordinators
-            if children_coordinators:
-                coordinator = children_coordinators[idx]
-            else:
-                coordinator = parent_coordinator
-            entities.extend(_async_sensors_for_device(child, coordinator, parent))
-    else:
-        entities.extend(_async_sensors_for_device(parent, parent_coordinator))
-
+    entities: list[CoordinatedTPLinkEntity] = []
+    device = parent_coordinator.device
+
+    entities = _entities_for_device_and_its_children(
+        device=device,
+        coordinator=parent_coordinator,
+        feature_type=Feature.Type.Sensor,
+        entity_class=TPLinkSensor,
+        child_coordinators=children_coordinators,
+    )
     async_add_entities(entities)
 
 
-class SmartPlugSensor(CoordinatedTPLinkEntity, SensorEntity):
-    """Representation of a TPLink Smart Plug energy sensor."""
+class TPLinkSensor(CoordinatedTPLinkEntity, SensorEntity):
```

Comment:
Would be cleaner to subclass `CoordinatedTPLinkEntity` as `CoordinatedTPLinkFeatureEntity` so `self._feature` is typed in the base class to always exist

Reply:
Done in https://github.com/home-assistant/core/pull/120025